Add registry support

jesusbv opened the pull request:
## Description

Add registry support to RMT

## Change Type

*Please select the correct option.*

- [ ] **Bug Fix** (a non-breaking change which fixes an issue)
- [X] **New Feature** (a non-breaking change which adds new functionality)
- [ ] **Documentation Update** (a change which only updates documentation)

## Checklist

*Please check off each item if the requirement is met.*

- [X] I have verified that my code follows RMT's coding standards with `rubocop`.
- [X] I have reviewed my own code and believe that it's ready for an external review.
- [ ] I have provided comments for any hard-to-understand code.
- [ ] I have documented the `MANUAL.md` file with any changes to the user experience.
- [ ] RMT's test coverage remains at 100%.
- [ ] If my changes are non-trivial, I have added a changelog entry to notify users at `package/obs/rmt-server.changes`.

## Other Notes

Please use this space to provide notes or thoughts to the team, such as tips on how to review/demo your changes.

rjschwei commented on engines/registry/README.md:
```diff
@@ -0,0 +1,3 @@
+# Registry
+Instances in the cloud have the option to use registry.suse.com for their containers,
+this engine provides the endpoint to auth and handling of that access to the registry
```
I don't get what this is trying to express and why registry.suse.com comes into play here. I suggest:
"""
Provide an authentication end point for a container registry. This supports access control based on system credentials to registry paths that are set to have access restrictions.
"""

rjschwei commented on engines/registry/app/controllers/registry/registry_controller.rb:
```diff
@@ -0,0 +1,92 @@
+module Registry
+  class RegistryController < ::ApplicationController
+    REGISTRY_SERVICE = 'SUSE Linux Docker Registry'.freeze
```
s/Docker/OCI/

bear454 commented on config/application.rb:
```diff
@@ -58,6 +58,15 @@ class Application < Rails::Application
 
+    # registry config needed
+    config.autoloader = :classic
+    registry_cert = "-----BEGIN PRIVATE KEY-----
```
Can we use the RMT server's private key here? The we can have the registry use the rmt-ca.crt for verification.

jesusbv commented on config/application.rb:
```diff
@@ -58,6 +58,15 @@ class Application < Rails::Application
+      File.read('/etc/rmt/ssl/rmt-server.key')
+      )
+    config.registry_public_key = config.registry_private_key.public_key
+    config.access_policies = '/etc/rmt/access_policies.yml'
```
The other option is to add the path to `/etc/rmt.conf`

rjschwei commented on config/application.rb:
```diff
@@ -58,6 +58,15 @@ class Application < Rails::Application
+    config.registry_private_key = OpenSSL::PKey::RSA.new(
+      File.read('/etc/rmt/ssl/rmt-server.key')
+      )
+    config.registry_public_key = config.registry_private_key.public_key
```
What are we doing with the key?

bear454 replied:
The OAuth token must be signed with a key, for which a certificate is available to the registry server, with the RMT server's FQDN as a SAN.

rjschwei commented on engines/registry/app/controllers/registry/registry_controller.rb:
```diff
@@ -0,0 +1,99 @@
+          @client = Registry::AuthenticatedClient.new(login, password)
+        rescue StandardError
+          logger.info _('Could not find system with login \"%{login}\" and password \"%{password}\"') %
+            { login: login, password: password }
```
his message does not help the user, it only tells the user that they are SOL. We need to tell the user what to do. In the special case we are currently implementing, i.e. for the public cloud update infrastructure only this should read.
"Registration expired please run cloudregistryauth command".  For air-gaped environments auth probably doesn't matter and our specialized message may be sufficient, but that requires more thought later. For now let's make sure we tell the user what they need to do.

digitaltom replied:
I'm don't think podman will show custom messages to the user at all. If I recall correctly it's showing it's own messages depending on the reponse code.

jesusbv replied:
It does, it would show 
```bash
> podman search registry-ec2.susecloud.net/
ERRO[0001] error getting search results from v2 endpoint "registry-ec2.susecloud.net": requested access to the resource is denied 
Error: 1 error occurred:
	* couldn't search registry "registry-ec2.susecloud.net": requested access to the resource is denied
```
that log information is for internal use, not for the user, happy to change the log message though

jesusbv replied:
@rjschwei what would `cloudregistryauth` command be ? 
We get the credentials when registering the instance on the client, 
I understand the command would be `registercloudguest` but, do we want the user
to re register the system to use (re-auth on) the registry ?

rjschwei commented on engines/registry/app/controllers/registry/registry_controller.rb:
```diff
@@ -0,0 +1,99 @@
+        begin
+          @client = Registry::CatalogClient.new(token)
+        rescue JWT::DecodeError
+          respond_with_error(message: 'Invalid token', status: :unauthorized) and return
```
We cannot simply tell the user that the token is invalid.

cwh42 replied:
We are replicating a OCI compliant registry API here and this is the expected response in case the token validation fails.
The client (docker/podman) will just get a new token, then. Usually no user action required.

rjschwei replied:
There will be no new token without re-authentication, i.e. we want/need to verify that the instance from which the request is coming from is still the instance that triggered the token generation when the system was registered. When the token expires, i.e. when we remove the system from the auth cache, then the token is expired until a re-authentication process happens. Since neither podman nor docker can send us the necessary data, see https://github.com/containers/podman/discussions/21697, we have to work around the limitation of those tools with our own code. Per agreement today that is going to be `cloudguestregistryauth`

rjschwei replied:
That said if the standard only allows for this response then that is the response we have to give and we have to figure out how to reduce the number of tickets generated via some other means

rjschwei commented on engines/registry/app/models/access_token.rb:
```diff
@@ -0,0 +1,46 @@
+class Registry::AccessToken
+  def initialize(account, service, granted_scopes)
+    @account = account
+    @service = service # "SUSE Linux Docker Registry"
```
s/Docker/OCI/

jesusbv commented on engines/registry/app/models/access_token.rb:
```diff
@@ -0,0 +1,46 @@
+class Registry::AccessToken
+  def initialize(account, service, granted_scopes)
+    @account = account
+    @service = service # "SUSE Linux OCI Registry"
```
Implies a change on the registry, as it currently has `Docker`

cwh42 replied:
Not sure if any of the clients does care about this.

rjschwei replied:
Docker is a TM and a couple of years ago we had an effort to remove the word "Docker" from everything that was not referring directly to the command itself.

rjschwei commented on engines/registry/app/controllers/registry/registry_controller.rb:
```diff
@@ -0,0 +1,100 @@
+        rescue StandardError
+          logger.info _('Could not find system with login \"%{login}\" and password \"%{password}\"') %
+            { login: login, password: password }
+          error = ActionController::TranslatedError.new(N_('Invalid registry credentials'))
```
I repeat we need to tell the user what to do. A generic message like "Invalid registry credentials" will do nothing but trigger the creation of support requests. We need to make at least an attempt to get the "you need to re-authenticate" message (different wording required) to the client. Making podman display it may be a follow up project in upstream podman if it doesn't display custom messages.

cwh42 commented on engines/registry/app/controllers/registry/registry_controller.rb:
```diff
@@ -0,0 +1,100 @@
+    end
+
+    # is called by authenticate_or_request_with_http_token when client provides no token
+    def request_http_token_authentication(realm = authorize_api_registry_url, message = 'authentication required')
```
Please have a look at the most recent version of [glue/app/controllers/api/registry_controller.rb](https://github.com/SUSE/happy-customer/blob/3f818723ea1b8cceab28e01505ea13d0570e28e1/glue/app/controllers/api/registry_controller.rb#L122
)
We had to add a fix here that broke some registry clients.

cwh42 commented on engines/registry/app/models/access_scope.rb:
```diff
@@ -0,0 +1,92 @@
+
+    allowed_products = (active_products & access_policies_yml.keys) + ['free']
+    allowed_glob_paths = access_policies_yml.values_at(*allowed_products).flatten
+    @allowed_paths = repo_list.select { |repo| repo if allowed_glob_paths.any? { |glob_path| File.fnmatch(glob_path, repo) } }
```
`File.fnmatch()` does not handle * and ** the same way as we defined it in SCC. So this does not give the right results!
For example `File.fnmatch('*')` matches *all* repos in the registry and not only those in the "root" directory as SCC does.

jesusbv replied:
mmmm, I had run some tests and the results seem OK ?

```
> podman search registry-ec2.susecloud.net/bci
NAME                                                 DESCRIPTION
registry-ec2.susecloud.net/opensuse/bci/bci-busybox  
registry-ec2.susecloud.net/opensuse/bci/ruby         
> podman search registry-ec2.susecloud.net/tool
NAME                                                                                                         DESCRIPTION
registry-ec2.susecloud.net/opensuse/leap/micro/5.2/images/totest/containers/opensuse/leap-micro/5.2/toolbox  
registry-ec2.susecloud.net/opensuse/leap/micro/5.3/images/totest/containers/opensuse/leap-micro/5.3/toolbox  
registry-ec2.susecloud.net/opensuse/leap/micro/5.4/images/totest/containers/opensuse/leap-micro/5.4/toolbox  
registry-ec2.susecloud.net/opensuse/leap/micro/5.5/images/images/opensuse/leap-micro/5.5/toolbox             
registry-ec2.susecloud.net/opensuse/leap/micro/5.5/images/totest/containers/opensuse/leap-micro/5.5/toolbox  
```
and no term/pattern returns _all_ free repos 
```
> podman search registry-ec2.susecloud.net/
NAME                                                                                                         DESCRIPTION
registry-ec2.susecloud.net/home/marcus.schaefer/delta_containers/containers_tw/fleet                         
registry-ec2.susecloud.net/home/marcus.schaefer/delta_containers/containers_tw/joe                           
registry-ec2.susecloud.net/opensuse/bci/bci-busybox                                                          
registry-ec2.susecloud.net/opensuse/bci/ruby                                                                 
registry-ec2.susecloud.net/opensuse/leap/micro/5.2/images/totest/containers/opensuse/leap-micro/5.2/toolbox  
registry-ec2.susecloud.net/opensuse/leap/micro/5.3/images/totest/containers/opensuse/leap-micro/5.3/toolbox  
registry-ec2.susecloud.net/opensuse/leap/micro/5.4/images/totest/containers/opensuse/leap-micro/5.4/toolbox  
registry-ec2.susecloud.net/opensuse/leap/micro/5.5/images/images/opensuse/leap-micro/5.5/toolbox             
registry-ec2.susecloud.net/opensuse/leap/micro/5.5/images/totest/containers/opensuse/leap-micro/5.5/toolbox  
> 
```

are those results not OK ?

digitaltom replied:
What @cwh42 meant is the meaning of * and ** in access policies: 

'*' applies to all images in the current path. For example, the policy `suse/*` applies to all images in the `suse/` directory, like `suse/389-ds`. But not to `suse/ltss/sle15.3/bci-base` for example. 

![image](https://github.com/SUSE/rmt/assets/582520/deafc702-5373-402f-a6f6-dc83efbeecaf)

'**' applies to all images and images in subdirectories of the path. Example for `suse/sles/**`: 

![image](https://github.com/SUSE/rmt/assets/582520/7b2101ad-a445-45ef-b4a4-2891eae17983)

jesusbv replied:
Updated

rjschwei commented on engines/registry/app/services/registry_catalog_service.rb:
```diff
@@ -0,0 +1,68 @@
+  end
+
+  def catalog_token(system)
+    system = System.where(proxy_byos: false).last if system.nil?
```
Why do we not allow systems that use the update server as a proxy? We want to allow any system that can register to the update server to access the registry if that is "allowed" according to what is included in a product.

jesusbv replied:
Good catch

That check does not ban access to BYOS system, we do allow them access but, for the catalog/search request, during testing, there was a request made without system, system being nil, and it crashed

That scenario is fixed (login, pull and search operations succeed) and the check should be removed

rjschwei commented on package/files/nginx-pubcloud/nginx-registry-https.conf:
```diff
@@ -0,0 +1,138 @@
+server {
+    listen 443 default_server  ssl;
+    listen [::]:443 default_server ssl;
+    server_name rmt;
```
But rmt, i.e. access to the repositories is covered in the existing config file `rmt-server-pubcloud-https.conf` why do we have to repeat the settings here?

rjschwei commented on package/files/nginx-pubcloud/nginx-registry-https.conf:
```diff
@@ -0,0 +1,138 @@
+server {
+    listen 443   ssl;
+    listen [::]:443 ssl;
+    server_name ~^registry;
```
Do we not need an auth setup for the registry path?

bear454 replied:
No, the auth is defined in the registry config.

bear454 commented:
@jesusbv we looked at two different approaches for the nginx config, but the current state reflects half of each.

* One https config for both rmt & registry, with the search endpoint of registry assigned as a location to rmt;
* Two separate configs, with the search endpoint in the registry config redirecting to rmt.

Which way did you intend to go?

bear454 commented on engines/registry/app/services/registry_catalog_service.rb:
```diff
@@ -36,8 +36,8 @@ def fetch_registry_repos(system)
   def catalog_token(system)
-    system = System.where(proxy_byos: false).last if system.nil?
-    uri = URI.parse(AUTH_URL)
+    # rmt.conf must have the root_url
```
Can we use the Request object to introspect the FQDN?

jesusbv commented:
> @jesusbv we looked at two different approaches for the nginx config, but the current state reflects half of each.
> 
>     * One https config for both rmt & registry, with the search endpoint of registry assigned as a location to rmt;
> 
>     * Two separate configs, with the search endpoint in the registry config redirecting to rmt.
> 
> 
> Which way did you intend to go?

The latter, split config, I think having the two configs separate is better, I updated the PR now to reflect that

Do you think one file is better ?

bear454 commented:
> The latter, split config, I think having the two configs separate is better, I updated the PR now to reflect that
> 
> Do you think one file is better ?

I think it's fine as long as it's been proven to work with both `podman` and `docker`.

rjschwei commented:
> > The latter, split config, I think having the two configs separate is better, I updated the PR now to reflect that
> > Do you think one file is better ?
> 
> I think it's fine as long as it's been proven to work with both `podman` and `docker`.

Separate config is preferable, that allows a 3rd party that doesn't want to run the registry to not install the config and then the paths are not available.

bear454 commented:
> > The latter, split config, I think having the two configs separate is better, I updated the PR now to reflect that
> > Do you think one file is better ?
> 
> I think it's fine as long as it's been proven to work with both `podman` and `docker`.

I'm withdrawing my objection here - as `docker` is incapable of searching via an authenticated endpoint, and only uses the unauthenticated /v1/search method. So, search of our registry will only be capable via `podman`.

bear454 commented:
> > > The latter, split config, I think having the two configs separate is better, I updated the PR now to reflect that
> > > Do you think one file is better ?
> > 
> > 
> > I think it's fine as long as it's been proven to work with both `podman` and `docker`.
> 
> Separate config is preferable, that allows a 3rd party that doesn't want to run the registry to not install the config and then the paths are not available.

Well, the problem is around search. Because we want to emulate the behavior of SCC, where the `/v2/_catalog` API endpoint is overriden, and served by _rmt_ instead of _distribution_, there is an intimate linking. The two-config solution requires that the registry config introduce a `302 REDIRECT` on that API endpoint, which also means the _rmt_ FQDN needs to be specified in the config.

With one config file, the `v2/_catalog`  API endpoint is overridden not to a `302 REDIRECT` but directly served by the defined _rmt_ application, as part of the reverse proxy configuration. The client doesn't see a redirect (that's not defined in the registry API) and get served the direct results on the original request. But yes, both _rmt_ and _registry_ vhosts have to be defined in the same file for that to work.

rjschwei commented:
> > > > The latter, split config, I think having the two configs separate is better, I updated the PR now to reflect that
> > > > Do you think one file is better ?
> > > 
> > > 
> > > I think it's fine as long as it's been proven to work with both `podman` and `docker`.
> > 
> > 
> > Separate config is preferable, that allows a 3rd party that doesn't want to run the registry to not install the config and then the paths are not available.
> 
> Well, the problem is around search. Because we want to emulate the behavior of SCC, where the `/v2/_catalog` API endpoint is overriden, and served by _rmt_ instead of _distribution_, there is an intimate linking. The two-config solution requires that the registry config introduce a `302 REDIRECT` on that API endpoint, which also means the _rmt_ FQDN needs to be specified in the config.
> 
> With one config file, the `v2/_catalog` API endpoint is overridden not to a `302 REDIRECT` but directly served by the defined _rmt_ application, as part of the reverse proxy configuration. The client doesn't see a redirect (that's not defined in the registry API) and get served the direct results on the original request. But yes, both _rmt_ and _registry_ vhosts have to be defined in the same file for that to work.

OK, thanks for the the details, then we are stuck with one config. But thta also means we should have 1 config, i.e. the existing config file should change rather than adding a new one.

bear454 commented on package/files/nginx-pubcloud/nginx-registry-https.conf:
```diff
@@ -0,0 +1,59 @@
+    # override registry search
+    location /v2/_catalog {
+      # redirect to the catalog endpoint
+      rewrite /v2/_catalog https://smt-ec2.susecloud.net/v2/_catalog redirect;
```
So, I woke up overnight and realized - this line is why we'll need to do the one-config approach. We don't want every RMT server in the wild redirecting to `smt-ec2.susecloud.net`... and there's no way to do the redirect without an FQDN.

jesusbv replied:
yes, hard coded string had to be fixed

felixsch commented on config/application.rb:
```diff
@@ -58,6 +58,19 @@ class Application < Rails::Application
     config.eager_load_paths << Rails.root.join('app', 'validators')
 
+    # :nocov:
+    if Rails.env.production?
```
I believe this needs to be gated behind a check if its actually running in public cloud mode, otherwise it will interfere with regular customers setups

jesusbv replied:
Added

rjschwei commented:
What's missing, I think, is the cache handling. We had agreed that clients that access the registry have a longer TTL than access to the repositories. Anyway, that change should come in a separate PR. This PR is big enough as it is and we have not made the desired timely progress in getting it merged.

bear454 commented:
> What's missing, I think, is the cache handling. We had agreed that clients that access the registry have a longer TTL than access to the repositories. Anyway, that change should come in a separate PR. This PR is big enough as it is and we have not made the desired timely progress in getting it merged.

Yes, agreed, this needs to be in a separate PR.

bear454 reviewed (approved):
There are still some magic strings in the code, but we can work on moving them out in subsequent PRs. That this is working, has a reliable config, and test coverage is good enough. Let's get rolling.

jesusbv commented:
> > What's missing, I think, is the cache handling. We had agreed that clients that access the registry have a longer TTL than access to the repositories. Anyway, that change should come in a separate PR. This PR is big enough as it is and we have not made the desired timely progress in getting it merged.
> 
> Yes, agreed, this needs to be in a separate PR.

Agree, yes, cache is not on this PR, and it should go on a different PR
